cli/coverage_cmd.go

```diff
@@ -0,0 +1,179 @@
+// The coverage heuristic is very rough currently, but makes the following assumptions:
+// - only golang coverage is checked
+// - if a directory can be imported (see build standard library package) and built
+//   (see loader standard library package), then there should be three files present
+//   in the corresponding directory under .srclib-cache: a unit file, a depresolve file,
+//   and a graph file.
+func (c *CoverageCmd) Execute(args []string) error {
+	if c.w == nil {
+		c.w = os.Stdout
+	}
+
+	lRepo, lRepoErr := OpenLocalRepo()
+	if lRepoErr != nil {
+		return lRepoErr
+	}
+
+	goPath := os.Getenv("GOPATH")
+	if goPath == "" {
+		return errors.New("GOPATH not set")
+	}
+
+	var importPath string
+	splitGoPath := filepath.SplitList(goPath)
+
+	for _, p := range splitGoPath {
+		p = filepath.ToSlash(p)
+		if strings.Contains(lRepo.RootDir, p) {
+			importPath = strings.TrimPrefix(lRepo.RootDir, filepath.Join(p, "src")+"/")
+		}
+	}
+
+	if importPath == "" {
+		return fmt.Errorf("Unable to find an import path for the current repo at %s", lRepo.RootDir)
+
+	}
+
+	cov := &Coverage{
+		Repo: lRepo,
+	}
+
+	_, err := os.Stat(filepath.Join(lRepo.RootDir, cacheDir))
+	if os.IsNotExist(err) {
+		return err
+	}
+
+	walker := fs.Walk(lRepo.RootDir)
+
+	for walker.Step() {
+
+		if err := walker.Err(); err != nil {
+			return err
+		}
+
+		pth := walker.Path()
+
+		fi, err := os.Stat(pth)
+		if !fi.IsDir() {
+			continue
+		}
+
+		if strings.Contains(pth, cacheDir) || strings.Contains(pth, gitDir) {
+			continue
+		}
+
+		relPath, err := filepath.Rel(lRepo.RootDir, pth)
+		if err != nil {
+			return err
+		}
+
+		_, importErr := build.ImportDir(pth, 0)
+
+		var conf loader.Config
+		conf.Import(strings.Join([]string{importPath, relPath}, "/"))
+		_, pkgErr := conf.Load()
+
+		importAndBuildSucceeded := (importErr == nil) && (pkgErr == nil)
+
+		// TODO(poler) allow the user to specify an older commit (fine for now)
+		cachePath := filepath.Join(lRepo.RootDir, cacheDir, lRepo.CommitID, importPath, relPath)
+
+		// If the srclib build config was at all customized, the assumption that these files
+		// will exist is almost certainly not valid.
+		unitPath := filepath.Join(cachePath, unitFile)
+		depPath := filepath.Join(cachePath, depFile)
+		graphPath := filepath.Join(cachePath, graphFile)
+
+		_, unitErr := os.Stat(unitPath)
+		_, depErr := os.Stat(depPath)
+		_, graphErr := os.Stat(graphPath)
+
+		srclibOutputsExist := (unitErr == nil) && (depErr == nil) && (graphErr == nil)
+
+		pth = filepath.ToSlash(pth)
+
+		if importAndBuildSucceeded && !srclibOutputsExist {
+			cov.Warnings = append(cov.Warnings, BuildWarning{
+				Directory: pth,
+				Warning:   BuildSucceededSrclibFailed,
+			})
+		} else if !importAndBuildSucceeded && srclibOutputsExist {
+			cov.Warnings = append(cov.Warnings, BuildWarning{
+				Directory: pth,
+				Warning:   BuildFailedSrclibSucceeded,
+			})
+		}
+	}
+
+	enc := json.NewEncoder(c.w)
```

Comment:
something else that would be useful in the json output would be some aggregate statistics on the warnings. Like number of Warnings, and a total for each warning type?